docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,153 @@
+the ability to execute whatever javascript they want. Any un-sanitized field that leaks HTML into your page is now an immediate vulnerability.
+
+There a great number of these vulnerable legacy endpoints, not just on googles domains but across the web.  
+Here is a list of [just some of them](https://github.com/zigoo0/JSONBee/blob/master/jsonp.txt).
+
+Because this isn't a specific protocol but just a serverside (anti)pattern, there's no way for the browser
+to detect this behavior and disable it.  The only solution is to not allow scripts to load from any vulnerable domain, and 
+to host them yourself instead.
+
+ **The bigger your website, the harder this is to deal with.**
+
+If you have a small site and immediate control over the codebase, this could be an easy problem to patch.
+
+For large sites at big companies that load analytics payloads, third party auth, and other scripts from these vulnerable domains, it can be 
+difficult to mitigate.  Heck, a lot of big companies have legacy JSONP endpoints on the _same domain as the site._  There's not much you can do about that.
+
+### File Upload Bypass
+Let's say your CSP is locked to only allow scripts from your domain.  
+
+If your site allows file uploads and downloads, including files that end in `.js`, a user can simply upload a file `malicious-code.js`,
+and load it as a script tag, directly from your domain. 
+
+```html
+<script src=www.yourdomain.com/myfiles/malicious-code.js></script>
+```
+
+You have to be very careful about what your users can upload.  
+
+### Other vulnerabilities
+These are just a few of the potential ways to bypass the CSP's protections.
+The list [goes on](https://book.hacktricks.xyz/pentesting-web/content-security-policy-csp-bypass).
+
+### Mitigations
+For large sites, there may be just too many chinks in the armor to treat the CSP as a functional line of defence, and many big sites have disabled it entirely. 
+[This paper](https://research.google/pubs/pub45542/) found that of those that use it **over 94% of sites have ineffective CSPs!**
+
+Long story short, there's a good chance your CSP isn't doing much to protect your site from XSS.
+
+Nonce-based CSP strategies can help mitigate the above attacks, where the server adds a random nonce attribute to every script tag and defines in the CSP.
+Embedding a script tag smart enough to clone the nonce isn't possible without javascript.  
+
+Unfortunately this strategy is the most work to implement and is rarely used. 
+
+```text title=CSP
+script-src 'nonce-randomString';
+```
+
+```html title="Script Tag"
+<script nonce="randomString" ...
+```
+
+Defining a nonce and a domain `script-src` together means that _either_ the domain or the nonce can be used, which seems like a
+gotcha that could catch many users.
+
+## How LunaSec and other tools use CSP effectively
+LunaSec is an Open Source security product we wrote, designed to help unhardened web-apps handle sensitive data. The main 
```

Comment:
Rewrite to, "[...] designed to harden web apps that handle sensitive data against leaking sensitive data, even _when_ CSP fails to protect you from XSS."

The word "unharden" is... the opposite of what we should be saying!

Reply:
also, `s/web-app/web app/` because that's how we have it everywhere else in our docs.